File: backend/app/api/v1/calendar.py
"""The `/api/v1/calendar` endpoint: the Department Leave Calendar — Manager-only (Story 3.3).

Implements: FR-18 / api-contracts §4.9 (`GET /calendar` | Role `Manager` | Scope `reports` —
a Manager's Direct Reports' Approved AND Pending leave overlapping a date range, presented at
the moment of decision, UJ-2), G3 (a non-Manager is refused `403 ACTION_NOT_PERMITTED` by the
role gate in the dependency, before any row is read), NFR-11 (the list is page-bounded),
AD-18 (`leave_days` on every item is the stored figure, never recomputed).

--- The contract inversion, second verse: the ADMIN is refused here ---

§4.9 grants `/calendar` to the Manager ALONE — exactly like `/team`: a calendar over a team is
a fact about a reporting edge, and only a Manager stands on one. An Admin reads any request via
`GET /leave-requests` (scope ALL); here `require_role(authz.ROLE_MANAGER)` refuses an Admin
exactly as it refuses an Employee.

--- Why there is NO `status` query param, and no status name in this module ---

FR-18 defines the calendar as "Approved and Pending" — the set is fixed SERVER-SIDE in
`services/calendar.py` (Open Decision #4), so the client cannot widen or narrow it and no
status name ever appears in `api/` (which may import neither `domain/` nor `repositories/` —
contract 2; the `Scope.REPORTS` decision lives in the service for the same reason).

--- Why the response shape is `LeaveRequestResponse`, reused byte-for-byte ---

Open Decision #1: a Manager already reads every one of these ten fields for these exact rows
via `GET /leave-requests` (scope REPORTS) — a narrower calendar shape would be a second
projection to maintain with zero disclosure gained, and `employee_name` is the field that
answers "who else is away." Imported from `leave_requests.py` with its projection (the
`DepartmentBrief`-from-`employees.py` single-home precedent); never redeclared.
"""

from __future__ import annotations

import datetime

from fastapi import APIRouter, Depends, Query

from app.api.v1.dependencies import Actor, require_role
from app.api.v1.leave_requests import LeaveRequestResponse, to_leave_request_response
from app.api.v1.pagination import Page, PageParams
from app.services import authorization as authz
from app.services import calendar as calendar_service

router = APIRouter()


@router.get("/calendar", tags=["calendar"])
def list_calendar(
    params: PageParams = Depends(),
    date_from: datetime.date | None = Query(default=None),
    date_to: datetime.date | None = Query(default=None),
    manager: Actor = Depends(require_role(authz.ROLE_MANAGER)),
) -> Page[LeaveRequestResponse]:
    """Return a page of the caller's Direct Reports' PENDING+APPROVED leave overlapping the
    window (AC1, AC2). Manager-only — an Employee AND an Admin are `403` (the §4.9 inversion).

    The role gate runs in the `manager` dependency, before this body — a non-Manager never
    reaches the read (G3). Both dates are optional (an absent side applies no predicate — never
    defaulted, so a window straddling Dec 31 works by construction); a malformed date is a
    framework `422` via the `datetime.date` typing; an inverted range is a well-formed empty
    intersection → `200` with `total == 0` (Landmine 5, the settled 3.1 semantics). The page is
    bounded by `PageParams` (NFR-11); the body carries the `items/page/page_size/total` envelope.
    """
    views, total = calendar_service.list_calendar(
        date_from=date_from,
        date_to=date_to,
        limit=params.limit,
        offset=params.offset,
        actor=manager,
    )
    return Page[LeaveRequestResponse](
        items=[to_leave_request_response(view) for view in views],
        page=params.page,
        page_size=params.page_size,
        total=total,
    )

File: frontend/src/api/calendar.ts
/**
 * The Department Leave Calendar, as a typed hook on `apiFetch`. `GET /api/v1/calendar` (Story 3.3).
 *
 * Implements: FR-18 (frontend) — a Manager's Direct Reports' PENDING + APPROVED leave overlapping
 * a date range, fetched at the moment of decision (UJ-2). Manager-ONLY server-side (api-contracts
 * §4.9, the same inversion as `/team`): an Employee AND an Admin get a `403`, so the caller gates
 * the fetch with `enabled` and the server's role gate stays the real guard. The status set is
 * fixed SERVER-side — there is no status param to send; Cancelled and Rejected leave is
 * deliberately not on a calendar.
 *
 * The wire shape IS `LeaveRequestResponse` (backend Open Decision #1) — the `LeaveRequest` type
 * from `./leaveRequests` is reused, never redeclared, exactly as `Page<T>` keeps its single home
 * in `departments.ts`. AD-2: every date, day count and status renders as received; nothing here
 * computes anything.
 */
import { useQuery } from '@tanstack/react-query'

import { apiFetch } from './client'
import type { Page } from './departments'
import type { LeaveRequest } from './leaveRequests'

/**
 * The base cache key for the calendar. Each params combination caches under
 * `[...CALENDAR_QUERY_KEY, params]` (TanStack v5 hashes keys structurally), so the decision
 * mutations' prefix invalidation (`invalidateAfterDecision` in `leaveRequests.ts`) reaches every
 * window at once — a decision flips a PENDING row to APPROVED/REJECTED, exactly the facts a
 * calendar under it is showing.
 */
export const CALENDAR_QUERY_KEY = ['calendar'] as const

/**
 * The calendar's params: the overlap window plus the shared page envelope's controls. All are
 * optional server-side (an absent date applies no predicate), but the decision calendar always
 * passes the request-under-decision's own range. Dates are `YYYY-MM-DD` strings rendered/sent
 * as received (AD-12/AD-2).
 */
export interface CalendarParams {
  dateFrom?: string
  dateTo?: string
  page?: number
  pageSize?: number
}

/** Each param's wire name, in one place so the query string and the backend stay in agreement. */
const PARAM_NAMES = {
  dateFrom: 'date_from',
  dateTo: 'date_to',
  page: 'page',
  pageSize: 'page_size',
} as const

/**
 * The Direct Reports' PENDING+APPROVED leave overlapping the window (FR-18). Every value is
 * `encodeURIComponent`-escaped into the query string (the 2.7 review's rule); the query key
 * carries the whole `params` object so each window caches distinctly while prefix invalidation
 * still fans out. `options.enabled` gates the fetch (the `useTeam` idiom): a non-Manager, which
 * renders nothing, never issues a request that is guaranteed to 403.
 */
export function useCalendar(
  params: CalendarParams = {},
  options?: { enabled?: boolean },
) {
  const pairs = (Object.keys(PARAM_NAMES) as (keyof CalendarParams)[])
    .filter((key) => params[key] !== undefined)
    .map((key) => `${PARAM_NAMES[key]}=${encodeURIComponent(String(params[key]))}`)
  const path: `/${string}` =
    pairs.length > 0 ? `/calendar?${pairs.join('&')}` : '/calendar'
  return useQuery({
    queryKey: [...CALENDAR_QUERY_KEY, params],
    queryFn: () => apiFetch<Page<LeaveRequest>>(path),
    enabled: options?.enabled,
  })
}
